src/webgpu/shader/validation/expression/binary/result_type.ts

```diff
@@ -0,0 +1,67 @@
+import {
+  isAbstractType,
+  isConvertible,
+  Type,
+  ScalarType,
+  VectorType,
+} from '../../../../util/conversion.js';
+
+/**
+ * @returns the resulting type of a binary arithmetic operation with the operands `lhs` and `rhs`
+ */
+export function resultType(args: {
+  lhs: Type;
+  rhs: Type;
+  canConvertScalarToVector: boolean;
+}): Type | null {
+  const { lhs, rhs, canConvertScalarToVector } = args;
+
+  if (lhs === rhs) {
+    return lhs;
+  }
+
+  if (lhs instanceof VectorType && rhs instanceof VectorType) {
+    // vector <op> vector
+    if (lhs.width !== rhs.width) {
+      return null;
+    }
+    const elementType = resultType({
+      lhs: lhs.elementType,
+      rhs: rhs.elementType,
+      canConvertScalarToVector,
+    }) as ScalarType | null;
+    return elementType !== null ? Type.vec(lhs.width, elementType) : null;
```

Comment:
I wonder if here and other places we can use the ?? operator (nullish coalescing). So it would be `return elementType ?? Type.vec(lhs.width, elementType)`

Reply:
That's not quite the same thing - we're returning a vector of `elementType`, if `elementType` is not `null`, not `elementType` itself.